embedded-graphics/src/primitives/rounded_rectangle.rs

```diff
@@ -0,0 +1,596 @@
+/// };
+///
+/// RoundedRectangle::new(Rectangle::new(Point::new(5, 5), Size::new(50, 60)), radii)
+///     .into_styled(style)
+///     .draw(&mut display)?;
+/// # Ok::<(), core::convert::Infallible>(())
+/// ```
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
+pub struct RoundedRectangle {
+    /// The base rectangle
+    pub rectangle: Rectangle,
+
+    /// The radius of each corner
+    pub corners: CornerRadii,
+}
+
+impl Primitive for RoundedRectangle {
+    type PointsIter = Points;
+
+    fn points(&self) -> Self::PointsIter {
+        Points::new(self)
+    }
+}
+
+impl ContainsPoint for RoundedRectangle {
+    fn contains(&self, point: Point) -> bool {
+        let Self {
+            rectangle, corners, ..
+        } = self;
+
+        let Rectangle { top_left, size, .. } = *rectangle;
+
+        if !rectangle.contains(point) {
+            return false;
+        }
+
+        let tl = EllipseQuadrant::new(top_left, corners.top_left, Quadrant::TopLeft);
+        let tr = EllipseQuadrant::new(
+            top_left + size.x_axis() - corners.top_right.x_axis(),
+            corners.top_right,
+            Quadrant::TopRight,
+        );
+        let br = EllipseQuadrant::new(
+            top_left + size - corners.bottom_right,
+            corners.bottom_right,
+            Quadrant::BottomRight,
+        );
+        let bl = EllipseQuadrant::new(
+            top_left + size.y_axis() - corners.bottom_left.y_axis(),
+            corners.bottom_left,
+            Quadrant::BottomLeft,
+        );
+
+        if tl.bounding_box().contains(point) {
+            return tl.contains(point);
+        }
+
+        if tr.bounding_box().contains(point) {
+            return tr.contains(point);
+        }
+
+        if br.bounding_box().contains(point) {
+            return br.contains(point);
+        }
+
+        if bl.bounding_box().contains(point) {
+            return bl.contains(point);
+        }
+
+        // We're in the rest of the rectangle at this point
+        true
+    }
+}
+
+impl Dimensions for RoundedRectangle {
+    fn bounding_box(&self) -> Rectangle {
+        self.rectangle
+    }
+}
+
+impl RoundedRectangle {
+    /// Creates a new rounded rectangle from a base rectangle and equal corner XY radius for  all
+    /// corners.
+    pub fn with_equal_corners(rectangle: Rectangle, corner_radius: Size) -> Self {
+        Self::new(rectangle, CornerRadii::new_equal(corner_radius))
+    }
+
+    /// Creates a new rounded rectangle with different corner radii.
+    pub fn new(rectangle: Rectangle, corners: CornerRadii) -> Self {
+        let Rectangle { size, .. } = rectangle;
+
+        let corners = CornerRadii {
+            top_left: corners.top_left.component_min(size / 2),
+            top_right: corners.top_right.component_min(size / 2),
+            bottom_right: corners.bottom_right.component_min(size / 2),
+            bottom_left: corners.bottom_left.component_min(size / 2),
```

Comment:
Ah, good catch. I could limit the sum of two corner radii to their total edge length perhaps? Do you have a better idea in mind?

Reply:
No, I don't have a better idea and i think limiting the sum is the right way to handle this. But the exact implementation how the radii are reduced to fit will require some consideration.

For reference: The way CSS handles this is described in: https://www.w3.org/TR/css-backgrounds-3/#corner-overlap. But this might require floating point math.

Reply:
I'll take a look at the CSS shrinking method later and see if it's doable with integer maths :crossed_fingers: